internal/xds/rbac/rbac_engine_test.go

```diff
@@ -1007,85 +1224,704 @@ func (s) TestChainEngine(t *testing.T) {
+					// address/port of this connection will be used to populate the
+					// destination ip/port in RPCData struct. This represents what
+					// the user of ChainEngine will have to place into
+					// context, as this is only way to get destination ip and port.
+					lis, err := net.Listen("tcp", "localhost:0")
+					if err != nil {
+						t.Fatalf("Error listening: %v", err)
+					}
+					defer lis.Close()
+					connCh := make(chan net.Conn, 1)
+					go func() {
+						conn, err := lis.Accept()
+						if err != nil {
+							t.Errorf("Error accepting connection: %v", err)
+							return
+						}
+						connCh <- conn
+					}()
+					_, err = net.Dial("tcp", lis.Addr().String())
+					if err != nil {
+						t.Fatalf("Error dialing: %v", err)
+					}
+					conn := <-connCh
+					defer conn.Close()
+					getConnection = func(context.Context) net.Conn {
+						return conn
+					}
+					ctx = peer.NewContext(ctx, data.rpcData.peerInfo)
+					stream := &ServerTransportStreamWithMethod{
+						method: data.rpcData.fullMethod,
+					}
+
+					ctx = grpc.NewContextWithServerTransportStream(ctx, stream)
+					err = cre.IsAuthorized(ctx)
+					if gotCode := status.Code(err); gotCode != data.wantStatusCode {
+						t.Fatalf("IsAuthorized(%+v, %+v) returned (%+v), want(%+v)", ctx, data.rpcData.fullMethod, gotCode, data.wantStatusCode)
+					}
+					if !reflect.DeepEqual(b.auditEvents, data.wantAuditEvents) {
+						t.Fatalf("Unexpected audit event for query:%v", data)
+					}
+
+					// this builder's auditEvents can be shared for several queries, make sure it's empty
+					b.auditEvents = nil
+				}()
+			}
+		})
+	}
+}
+
+type ServerTransportStreamWithMethod struct {
+	method string
+}
+
+func (sts *ServerTransportStreamWithMethod) Method() string {
+	return sts.method
+}
+
+func (sts *ServerTransportStreamWithMethod) SetHeader(md metadata.MD) error {
+	return nil
+}
+
+func (sts *ServerTransportStreamWithMethod) SendHeader(md metadata.MD) error {
+	return nil
+}
+
+func (sts *ServerTransportStreamWithMethod) SetTrailer(md metadata.MD) error {
+	return nil
+}
+
+// An audit logger that will log to the auditEvents slice
+type TestAuditLoggerBuffer struct {
+	auditEvents *[]*audit.Event
+}
+
+func (logger *TestAuditLoggerBuffer) Log(e *audit.Event) {
+	*(logger.auditEvents) = append(*(logger.auditEvents), e)
+}
+
+// Builds TestAuditLoggerBuffer
+type TestAuditLoggerBufferBuilder struct {
+	auditEvents []*audit.Event
+}
+
+// The required config for TestAuditLoggerBuffer
+type TestAuditLoggerBufferConfig struct {
+	audit.LoggerConfig
+}
+
+func (builder *TestAuditLoggerBufferBuilder) ParseLoggerConfig(configJSON json.RawMessage) (config audit.LoggerConfig, err error) {
+	return TestAuditLoggerBufferConfig{}, nil
+}
+
+func (builder *TestAuditLoggerBufferBuilder) Build(config audit.LoggerConfig) audit.Logger {
+	return &TestAuditLoggerBuffer{auditEvents: &builder.auditEvents}
+}
+
+func (builder *TestAuditLoggerBufferBuilder) Name() string {
```

Comment:
It seems to me like, we use audit logger `Name()` for registry. How about we make `TestAuditLoggerBufferBuilder` a little smarter to return a unique name for each test case. Im thinking along lines of appending maybe the testcase.name in each scenario. That way we dont need to unregister the entry by having unique names for each test case. 

thoughts?

Reply:
Discussing in other comment thread